src/core/qgsuserprofilemanager.cpp

```diff
@@ -0,0 +1,200 @@
+{
+  return !mRootProfilePath.isEmpty();
+}
+
+QString QgsUserProfileManager::defaultProfileName() const
+{
+  QString profileName = "default";
+  mSettings->value( "/defaultProfile", "default" );
+  return profileName;
+}
+
+void QgsUserProfileManager::setDefaultProfileName( const QString &name )
+{
+  mSettings->setValue( "/defaultProfile", name );
+  mSettings->sync();
+}
+
+void QgsUserProfileManager::setDefaultFromActive()
+{
+  setDefaultProfileName( userProfile()->name() );
+}
+
+QStringList QgsUserProfileManager::allProfiles() const
+{
+  return QDir( mRootProfilePath ).entryList( QDir::Dirs | QDir::NoDotAndDotDot );
+}
+
+bool QgsUserProfileManager::profileExists( const QString &name ) const
+{
+  return allProfiles().contains( name );
+}
+
+QgsUserProfile *QgsUserProfileManager::profileForName( const QString name ) const
+{
+  QString profilePath = mRootProfilePath + QDir::separator() + name;
+  return new QgsUserProfile( profilePath );
+}
+
+QgsError QgsUserProfileManager::createUserProfile( const QString &name )
+{
+  QgsError error;
+
+  // TODO Replace with safe folder name
+
+  QDir folder( mRootProfilePath + QDir::separator() + name );
+  if ( !folder.exists() )
+  {
+    QDir().mkdir( folder.absolutePath() );
+  }
+
+  QFile qgisPrivateDbFile( folder.absolutePath() + QDir::separator() + "qgis.db" );
+
+  // first we look for ~/.qgis/qgis.db
+  if ( !qgisPrivateDbFile.exists() )
+  {
+    // if it doesn't exist we copy it in from the global resources dir
+    QString qgisMasterDbFileName = QgsApplication::qgisMasterDatabaseFilePath();
+    QFile masterFile( qgisMasterDbFileName );
+
+    //now copy the master file into the users .qgis dir
+    masterFile.copy( qgisPrivateDbFile.fileName() );
+  }
+
+  if ( error.isEmpty() )
+  {
+    emit profilesChanged();
+  }
+
+  return error;
+}
+
+QgsError QgsUserProfileManager::deleteProfile( const QString name )
+{
+  QgsError error;
+  QDir folder( mRootProfilePath + QDir::separator() + name );
+
+  // This might have be changed to something better.
+  bool deleted = folder.removeRecursively();
+  if ( !deleted )
+  {
+    error.append( ( tr( "Unable to fully delete user profile folder" ) ) );
+  }
+  else
+  {
+    emit profilesChanged();
+  }
+  return error;
+}
+
+QString QgsUserProfileManager::settingsFile()
+{
+  return  mRootProfilePath + QDir::separator() + "profiles.ini";
+}
+
+QgsUserProfile *QgsUserProfileManager::userProfile()
+{
+  return mUserProfile.get();
+}
+
+void QgsUserProfileManager::loadUserProfile( const QString &name )
+{
+  QString path = QDir::toNativeSeparators( QCoreApplication::applicationFilePath() );
+  QStringList arguments;
+  arguments << QCoreApplication::arguments();
+  // The first is the path to the application
+  // on Windows this might not be case so we need to handle that
+  // http://doc.qt.io/qt-5/qcoreapplication.html#arguments
+  arguments.removeFirst();
+
+  arguments << "--profile" << name;
```

Comment:
I've not tested, but what happens here if qgis was already launched with `--profile` ? Does the second override the first? If this is not the case, a more robust approach would be required to parse the arguments.